### PR TITLE
Fix: IT in GitLab dev failed due to empty manifests (#4392)

### PR DESCRIPTION
**Pull request by achave11**
Connected issue: #0000

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                   <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased PR branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed PR branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label    <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment                          <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev`                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `anvilbox`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `anvilbox`                                 <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox`                            <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [ ] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [ ] Pushed merge commit to GitLab `dev`                           <sub>or PR is labeled `no sandbox`</sub>
- [ ] Pushed merge commit to GitLab `anvildev`                      <sub>or PR is labeled `no sandbox`</sub>
- [ ] Build passes on GitLab `dev`<sup>1</sup>
- [ ] Build passes on GitLab `anvildev`<sup>1</sup>
- [ ] Deleted PR branch from GitHub
- [ ] Deleted PR branch from GitLab `dev`
- [ ] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is* 
pushed determines this checklist item.

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Deleted unreferenced indices in `anvildev`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Started reindex in `anvildev`                                 <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures in `dev`            <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures in `anvildev`       <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in `dev` deployment                       <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in `anvildev` deployment                  <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by achave11** [comment]
It was pointless to attempt to reproduce this in sandbox because sandbox has a prefix and `dev` doesn't. Reproduce locally, with `dev` selected and pin the seed.

**Comment by achave11** [comment]
Repro attained locally with `$_select dev` while running the IT with the pinned seed.
```
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest) [managed_access_manifest]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/.venv/lib/python3.9/site-packages/more_itertools/more.py", line 138, in first
    return next(iter(iterable))
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 370, in subTest
    yield
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 1027, in _test_managed_access
    first(public_source_ids & indexed_source_ids))
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/.venv/lib/python3.9/site-packages/more_itertools/more.py", line 145, in first
    raise ValueError('first() was called on an empty iterable, and no '
ValueError: first() was called on an empty iterable, and no default value was provided.

======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest) [manifest] (catalog='dcp2-it', format=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 370, in subTest
    yield
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 535, in _test_manifest
    validator(catalog, response)
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 645, in _check_manifest
    self.__check_manifest(BytesIO(response), 'bundle_uuid')
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 721, in __check_manifest
    self.assertGreater(len(rows), 0)
AssertionError: 0 not greater than 0

======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest) [manifest] (catalog='dcp2-it', format=<ManifestFormat.compact: 'compact'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 370, in subTest
    yield
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 535, in _test_manifest
    validator(catalog, response)
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 645, in _check_manifest
    self.__check_manifest(BytesIO(response), 'bundle_uuid')
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 721, in __check_manifest
    self.assertGreater(len(rows), 0)
AssertionError: 0 not greater than 0

======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest) [manifest] (catalog='dcp2-it', format=<ManifestFormat.terra_bdbag: 'terra.bdbag'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 370, in subTest
    yield
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 535, in _test_manifest
    validator(catalog, response)
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 652, in _check_terra_bdbag
    rows = self.__check_manifest(file, 'bundle_uuid')
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 721, in __check_manifest
    self.assertGreater(len(rows), 0)
AssertionError: 0 not greater than 0

======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest) [manifest] (catalog='dcp2-it', format=<ManifestFormat.curl: 'curl'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 370, in subTest
    yield
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 535, in _test_manifest
    validator(catalog, response)
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 745, in _check_curl_manifest
    self.assertGreater(num_files, 0)
AssertionError: 0 not greater than 0

======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest) [repository_files] (catalog='dcp2-it')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 370, in subTest
    yield
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 749, in _test_repository_files
    file = self._get_one_file(catalog)
  File "/Users/abrahamchavez/AbrahamC/azul.release/azul.stable/test/integration_test.py", line 551, in _get_one_file
    self.fail('No files found')
AssertionError: No files found

----------------------------------------------------------------------
Ran 16 tests in 687.879s

FAILED (failures=5, errors=1, skipped=6)
make: *** [integration_test] Error 1

```

File: test/integration_test.py [hunk 1]
```diff
@@ -314,9 +314,20 @@ def _list_bundles(self,
...
         total_bundles = 0
-        sources = sorted(config.sources(catalog))
-        self.random.shuffle(sources)
+        sources = config.sources(catalog)
+        managed_access_sources = self.managed_access_sources_by_catalog[catalog]
+        public_sources = sources - frozenset(map(str, managed_access_sources))
+        sources = sorted(sources)
+        while True:
+            self.random.shuffle(sources)
+            # Ensure that at least one public source will be indexed
```
Comment (danielsotirhos) [inline test/integration_test.py:325]:
```suggestion
```

File: test/integration_test.py [hunk 2]
```diff
@@ -314,9 +314,20 @@ def _list_bundles(self,
...
+            if sources[0] in public_sources:
+                break
+            elif not public_sources:
+                log.warning('No public sources found in catalog %r', catalog)
+                break
         # This iteration prefers sources occurring first, so we shuffle them
         # above to neutralize the bias.
```
Comment (danielsotirhos) [inline test/integration_test.py:332]:
```suggestion
        # This iteration prefers sources occurring first, so we shuffle them
        # above to neutralize the bias, making sure the sequence starts with
        # at least one public source if any exist in the catalog.
```

**Review by danielsotirhos — CHANGES_REQUESTED** [review]
Subjective suggestions on the wording / placement of the comments. Feel free to push back if you think your version reads better.

**Comment by noah-aviel-dove** [comment]
Decided to take a different approach. Hope this is clearer.

**Review by danielsotirhos — APPROVED** [review]
Approved.

File: test/integration_test.py [hunk 3]
```diff
@@ -904,12 +914,14 @@ def update(source: SourceRef,
                                                                   prefix))
 
         list(starmap(update, self._list_managed_access_bundles(catalog)))
-        num_bundles = self.min_bundles - len(bundle_fqids)
+        # Ensure that at least one public source will be indexed
```
Comment (hannes-ucsc) [inline test/integration_test.py:917]: This comment seems a little far away from the actual measure that's being taken, which is passing `public_1st=True`. It also should explain why the measure is taken. I usually explain what would happen if I didn't take the measure.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/integration_test.py [hunk 4]
```diff
@@ -904,12 +914,19 @@ def update(source: SourceRef,
...
 
         list(starmap(update, self._list_managed_access_bundles(catalog)))
-        num_bundles = self.min_bundles - len(bundle_fqids)
+        num_bundles = max(self.min_bundles - len(bundle_fqids), 1)
         log.info('Selected %d bundles to satisfy managed access coverage; '
                  'selecting at least %d more', len(bundle_fqids), num_bundles)
+        # _list_bundles selects both public and managed access sources at random.
+        # If we don't index at least one public source, every request would need
+        # service account credentials and we couldn't compare the responses for
+        # public and managed access data. `public_first` ensures that at least
```
Comment (hannes-ucsc) [inline test/integration_test.py:923]:
```suggestion
        # public and managed access data. `public_1st` ensures that at least
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Pull request merged**